src/hotkey/hotkey_item.cpp

```diff
@@ -24,537 +25,471 @@
+	item["ctrl"] = !!(mod_ & KMOD_CTRL);
+#if SDL_VERSION_ATLEAST(2, 0, 0)
+	item["cmd"] = !!(mod_ & KMOD_GUI);
+#else
+	item["cmd"] = !!(mod_ & KMOD_META);
+#endif
+	item["alt"] = !!(mod_ & KMOD_ALT);
 
+	save_helper(item);
+}
 
-const hotkey_item& get_hotkey(const SDL_KeyboardEvent& event)
+hotkey_ptr create_hotkey(const std::string &id, SDL_Event &event)
 {
-	return get_hotkey(
-#if SDL_VERSION_ATLEAST(2, 0, 0)
-			event.keysym.scancode,
+	hotkey_ptr base = hotkey_ptr(new hotkey_void);
+
+	switch (event.type) {
+	case SDL_KEYDOWN:
+	case SDL_KEYUP: {
+		hotkey_keyboard_ptr keyboard(new hotkey_keyboard());
+		base = boost::dynamic_pointer_cast<hotkey_base>(keyboard);
+		SDL_Scancode code;
+#if  SDL_VERSION_ATLEAST(2, 0, 0)
+		code = event.key.keysym.scancode;
 #else
-			event.keysym.unicode,
+		code = SDL_GetScancodeFromKey(event.key.keysym.sym);
 #endif
-			event.keysym.sym,
-			(event.keysym.mod & KMOD_SHIFT) != 0,
-			(event.keysym.mod & KMOD_CTRL)  != 0,
-			(event.keysym.mod & KMOD_META)  != 0,
-			(event.keysym.mod & KMOD_ALT)   != 0
-			);
+		keyboard->set_scancode(code);
+		break;
+	}
+	case SDL_MOUSEBUTTONDOWN:
+	case SDL_MOUSEBUTTONUP: {
+		hotkey_mouse_ptr mouse(new hotkey_mouse());
+		base = boost::dynamic_pointer_cast<hotkey_base>(mouse);
+		mouse->set_button(event.button.button);
+		break;
+	}
+	default:
+		ERR_G<< "Trying to bind an unknown event type:" << event.type << "\n";
+		break;
+	}
+
+	base->set_mods(sdl_get_mods());
+	base->set_command(id);
+	base->unset_default();
+
+	return base;
 }
 
+hotkey_ptr load_from_config(const config& cfg)
+{
+	hotkey_ptr base = hotkey_ptr(new hotkey_void());
+
+	const std::string& mouse_cfg = cfg["mouse"];
+	if (!mouse_cfg.empty()) {
+		hotkey_mouse_ptr mouse(new hotkey_mouse());
+		base = boost::dynamic_pointer_cast<hotkey_base>(mouse);
+		mouse->set_button(cfg["button"].to_int());
+	}
+	// TODO: add joystick support back
+#if 0
+	const std::string& joystick_cfg = cfg["joystick"];
+	if (!joystick_cfg.empty()) {
+		joystick_ = cfg["joystick"].to_int();
+	}
+	const std::string& hat = cfg["hat"];
+	if (!hat.empty()) {
+		hat_ = cfg["hat"].to_int();
+		value_ = cfg["value"].to_int();
+	}
 
+	const std::string& button = cfg["button"];
+	if (!button.empty()) {
+		button_ = cfg["button"].to_int();
+	}
+#endif
 
-std::string get_names(std::string id) {
+	const std::string& key_cfg = cfg["key"];
+	if (!key_cfg.empty()) {
+		hotkey_keyboard_ptr keyboard(new hotkey_keyboard());
+		base = boost::dynamic_pointer_cast<hotkey_base>(keyboard);
 
-	std::vector<std::string> names;
-	BOOST_FOREACH(const hotkey::hotkey_item& item, hotkeys_) {
-		if (item.get_command() == id && (!item.null()) ) {
-			names.push_back(item.get_name());
+		SDL_Scancode scancode = SDL_GetScancodeFromName(key_cfg.c_str());
+		if (scancode == SDL_SCANCODE_UNKNOWN) {
+			ERR_G<< "Unknown key: " << key_cfg << "\n";
 		}
+		keyboard->set_scancode(scancode);
+	}
+	
+	// These are hard-coded, non-rebindable hotkeys
+	if(id == "quit") {
```

Comment:
I think something horrible happened here when you rebased. Is this snippet of code supposed to be in get_names() instead? There should also be a space between the if and the opening bracket to fit into the style in the rest of the file.